test/TorchSharpTest/TestTraining.cs

```diff
@@ -99,8 +101,10 @@ public void TestTrainingWithDropout()
                 using (torch.no_grad()) {
                     foreach (var param in seq.parameters()) {
                         var grad = param.grad();
-                        var update = grad.mul(learning_rate);
-                        param.sub_(update);
+                        if (grad is not null) {
```

Comment:
I wonder if simply raising a NullReferenceException may be better than requiring the nullness checks, in the balance of things. It's in the nature of tensor programming to have failure points and it's not necessarily what we're trying to solve here...

Reply:
Except that grad() being null is not a failure -- it is expected to be null before backward() is called the first time, as well as when you are writing optimizers (that's how this came up), when you often set 'no_grad()' for a while. In PyTorch, there's also an option to set grad to null rather than zeroing when you call zero_grad() (which TorchSharp doesn't implement, but may in the future). Anyway, an exception would be bad, IMO.